src/pdf_beamtime/src/pdf_beamtime_fidpose_server.cpp

```diff
@@ -0,0 +1,626 @@
+    std::bind(&PdfBeamtimeFidPoseServer::fidpose_handle_accepted, this, _1));
+}
+
+rclcpp_action::GoalResponse PdfBeamtimeFidPoseServer::fidpose_handle_goal(
+  const rclcpp_action::GoalUUID & uuid,
+  std::shared_ptr<const FidPoseControlMsg::Goal> goal)
+{
+  (void)uuid;
+  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+void PdfBeamtimeFidPoseServer::fidpose_handle_accepted(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<FidPoseControlMsg>> goal_handle)
+{
+  using namespace std::placeholders;
+
+  // this needs to return quickly to avoid blocking the executor, so spin up a new thread
+  std::thread{std::bind(&PdfBeamtimeFidPoseServer::execute, this, _1), goal_handle}.detach();
+}
+
+// Receiving the cancel request.
+rclcpp_action::CancelResponse PdfBeamtimeFidPoseServer::fidpose_handle_cancel(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<FidPoseControlMsg>> goal_handle)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to cancel goal");
+  (void)goal_handle;
+  return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+void PdfBeamtimeFidPoseServer::execute(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<FidPoseControlMsg>> goal_handle)
+{
+  fidpose_goal = goal_handle->get_goal();
+  //  Variables for feedback and results
+  auto feedback = std::make_shared<FidPoseControlMsg::Feedback>();
+  auto results = std::make_shared<FidPoseControlMsg::Result>();
+  moveit::core::MoveItErrorCode fsm_results;
+  progress_ = 0.0;
+  goal_home_ = node_->get_parameter("home_angles").as_double_array();
+  pickup_approach_ = node_->get_parameter("pickup_approach_angles").as_double_array();
+  feedback->status = get_action_completion_percentage();
+
+  RCLCPP_INFO(
+    node_->get_logger(), "Current state is %s.",
+    external_state_names_[static_cast<int>(current_state_)].c_str());
+
+  // Reset inner_state_machine at new goal
+  inner_state_machine_->set_internal_state(Internal_State::RESTING);
+
+  // Keep executing the states until the a goal is completed or cancelled
+  while (get_action_completion_percentage() < 99.99999) {
+    if (inner_state_machine_->get_internal_state() == Internal_State::PAUSED) {
+      // Upon triggering pause_, the execution while loop switches to 1HZ.
+      // The external and internal states are handled separately by the FSM
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    } else {
+      if (inner_state_machine_->get_internal_state() == Internal_State::CLEANUP) {
+        // Abort the goal and return at cleanup
+        results->success = false;
+        goal_handle->abort(results);
+        RCLCPP_ERROR(node_->get_logger(), "Goal aborted !");
+        return;
+      }
+      if (fidpose_goal->sample_return) {
+        if (pickup_pose_saved) {
+          fsm_results = run_return_fsm(fidpose_goal);
+        } else {
+          RCLCPP_ERROR(node_->get_logger(), "Pickup poses are not set!");
+          return;
+        }
```

Comment:
It's fine to leave this error logic as a failsafe, but shouldn't the goal itself should be rejected if its not a feasible goal?

Reply:
Resolved in the earlier comment.